Pickling objects with __slots__ fails

The following minimal test case:

```diff
diff --git a/tests/cloudpickle_test.py b/tests/cloudpickle_test.py
index d318da7..fec779c 100644
--- a/tests/cloudpickle_test.py
+++ b/tests/cloudpickle_test.py
@@ -1310,6 +1310,17 @@ class CloudPickleTest(unittest.TestCase):
 
         self.assertEqual(f2.__annotations__, f.__annotations__)
 
+    def test_slots(self):
+        class WithSlots(object):
+            __slots__ = ["a"]
+
+            def __init__(self):
+                self.a = 42
+
+        obj = WithSlots()
+        obj2 = pickle_depickle(obj, protocol=self.protocol)
+        self.assertEqual(obj2.a, 42)
+
 
 class Protocol2CloudPickleTest(CloudPickleTest):

```

Fails with this exception:

```
obj = <tests.cloudpickle_test.CloudPickleTest.test_slots.<locals>.WithSlots object at 0x7f5cdc8f94c8>, protocol = 4

    def pickle_depickle(obj, protocol=cloudpickle.DEFAULT_PROTOCOL):
        """Helper function to test whether object pickled with cloudpickle can be
        depickled with pickle
        """
>       return pickle.loads(cloudpickle.dumps(obj, protocol=protocol))
E       AttributeError: type object 'WithSlots' has no attribute 'a'

obj        = <tests.cloudpickle_test.CloudPickleTest.test_slots.<locals>.WithSlots object at 0x7f5cdc8f94c8>
protocol   = 4
```

Pickling with `pickle` from the standard library works.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,3 +1,11 @@
+0.7.0
+=====
+
+- Correctly serialize dynamically defined classes that have a `__slots__`
+  attribute.
+  ([issue #225](https://github.com/cloudpipe/cloudpickle/issues/225))
+
+
 0.6.1
 =====
 

--- a/cloudpickle/cloudpickle.py
+++ b/cloudpickle/cloudpickle.py
@@ -55,6 +55,7 @@ import operator
 import importlib
 import itertools
 import traceback
+import six
 from functools import partial
 
 
@@ -499,6 +500,17 @@ class CloudPickler(Pickler):
         # the initial skeleton class.  This is safe because we know that the
         # doc can't participate in a cycle with the original class.
         type_kwargs = {'__doc__': clsdict.pop('__doc__', None)}
+
+        if hasattr(obj, "__slots__"):
+            type_kwargs['__slots__'] = obj.__slots__
+            # pickle string length optimization: member descriptors of obj are
+            # created automatically from obj's __slots__ attribute, no need to
+            # save them in obj's state
+            if isinstance(obj.__slots__, six.string_types):
+                clsdict.pop(obj.__slots__)
+            else:
+                for k in obj.__slots__:
+                    clsdict.pop(k, None)
 
         # If type overrides __dict__ as a property, include it in the type kwargs.
         # In Python 2, we can't set this attribute after construction.

--- a/tests/cloudpickle_test.py
+++ b/tests/cloudpickle_test.py
@@ -1310,6 +1310,23 @@ class CloudPickleTest(unittest.TestCase):
 
         self.assertEqual(f2.__annotations__, f.__annotations__)
 
+    def test_instance_with_slots(self):
+        for slots in [["registered_attribute"], "registered_attribute"]:
+            class ClassWithSlots(object):
+                __slots__ = slots
+
+                def __init__(self):
+                    self.registered_attribute = 42
+
+            initial_obj = ClassWithSlots()
+            depickled_obj = pickle_depickle(
+                initial_obj, protocol=self.protocol)
+
+            for obj in [initial_obj, depickled_obj]:
+                self.assertEqual(obj.registered_attribute, 42)
+                with pytest.raises(AttributeError):
+                    obj.non_registered_attribute = 1
+
 
 class Protocol2CloudPickleTest(CloudPickleTest):
 

